### DISCUSSION TITLE
Generate pact files for the mocks that can be published to a pact broker

### DISCUSSION THREAD

**Discussion by skirankumars31** [microcks/microcks#803] [Ideas]
Hi,

A lot of companies follow contract testing as a standard. It would be great if the mocks generated by microcks can generate pact contract files that can be tested against the real service via a pact broker. This idea is same as using the pact-wiremock-generator

https://bitbucket.org/atlassian/wiremock-pact-generator

Wiremock-pact-generator will generate pact files for the mocks exposed by wiremock. Why not we use the same model ?

Regards
Kiran

**Comment by lbroudoux** [comment]
Hey Kiran,

The question of similarities/differences between Pact and Microcks is raised more and more these last weeks. Not a Pact expert but currently diving in it. From your point of view: what would be the differences and benefits of generating pact contract versus using directly Microcks for this?

From my early findings it seems to cover exactly the same feature set: with Microcks tests you're replaying your samples against the real services and then validate the retrieved responses are conformant to the contract (be it OpenAPI, AsyncAPI, GRPC, GraphQL, SOAP, Postman assertions, ...).

I really like to learn more on Pact and how we complement/overlap it.

**Comment by skirankumars31** [comment]
1. Using a pact broker will give the provider a better visibiity of which mocks are conformant to their specification and which are not.
2.  I think the new pact spec 3 support async events too.
3.  Where there are several versions of the consumers and provider involved, pact would be a much simpler way.

**Reply by lbroudoux** [reply]
Thanks for the 3 points. Still have to read much more on Pact to be sure I grasp everything... However here are some of my thoughts and questions at the moment:

1/ This is a point we could enhance with Microcks as of today we do not validate that the examples contained in artifacts are valid regarding the schema. DO you think this would help to have this validation at the time we import a new spec? 

2/ From what I've read so far, it supports contract testing event structures but as Pact is dedicated to unit tests it doesn't deal with protocol specific or serialization concerns. Saw it in https://docs.pact.io/getting_started/how_pact_works#non-http-testing-message-pact where they talk about avoiding testing the _Adapter_ part. Do you have some other pointers?

3/ Could you explicit that point? As we're having different versions with different endpoints in Microcks too + the ability to test an endpoint against multiple versions of same schemas, I wonder how Pact would make things simpler?

Really want to understand this! Thanks,
